bpf/tpinjector/tpinjector.c

```diff
@@ -22,59 +26,180 @@
 
 #include <logger/bpf_dbg.h>
 
+#include <maps/incoming_trace_map.h>
 #include <maps/msg_buffers.h>
 #include <maps/ongoing_http.h>
 #include <maps/sock_dir.h>
+#include <maps/sock_pids.h>
 
-#include <tpinjector/maps/egress_key_mem.h>
 #include <tpinjector/maps/extender_jump_table.h>
 #include <tpinjector/maps/pid_connection_info_mem.h>
+#include <tpinjector/maps/sk_tp_info_pid_map.h>
 
 char __license[] SEC("license") = "Dual MIT/GPL";
 
+// Flags to control what tpinjector should inject
+enum {
+    k_inject_http_headers = 1 << 0, // Bit 0: inject HTTP headers
+    k_inject_tcp_options = 1 << 1,  // Bit 1: inject TCP options
+};
+
+volatile const u32 inject_flags =
+    k_inject_http_headers | k_inject_tcp_options; // default: both enabled
+
+// TCP option kind for OpenTelemetry context propagation
+// Kind 25 is unassigned per IANA TCP Parameters registry (released 2000-12-18)
+// Better than experimental options (253-254) which must not be shipped as defaults
+enum { k_tcp_option_kind_otel = 25 };
+
 enum { k_tail_write_msg_traceparent = 0 };
 
+SCRATCH_MEM_SIZED(tp_str_buf, 64)
+
+#ifndef ENOMSG
+#define ENOMSG 42
+#endif
+
+struct tp_option {
+    u8 kind;
+    u8 len;
+    unsigned char trace_id[TRACE_ID_SIZE_BYTES];
+    unsigned char span_id[SPAN_ID_SIZE_BYTES];
+};
+
+static __always_inline void
+encode_hex_skb(unsigned char *dst, const unsigned char *src, u32 src_len) {
+
+#pragma clang loop unroll(full)
+    for (u32 i = 0, j = 0; i < src_len; i++) {
+        unsigned char p = src[i];
+
+        dst[j++] = hex[(p >> 4) & 0xff];
+        dst[j++] = hex[p & 0x0f];
+    }
+}
+
+static __always_inline const char *tp_string_from_opt(const struct tp_option *opt) {
```

Comment:
we also have something similar in `common/tracing.h` (see: `make_tp_string()`), maybe we can use the tp_info instead of tp_option